libafl_frida/src/cmplog_rt.rs

```diff
@@ -243,28 +309,261 @@ impl CmpLogRuntime {
+        tmp_reg.insert(2, Register::AX);
+
+        #[cfg(windows)]
+        {
+            arg_reg_1 = Register::CL;
+            arg_reg_2 = Register::RDX;
+            arg_reg_3 = Register::R8;
+            arg_reg_4 = Register::R9;
+        }
+        #[cfg(unix)]
+        {
+            arg_reg_1 = Register::DL;
+            arg_reg_2 = Register::RSI;
+            arg_reg_3 = Register::RDX;
+            arg_reg_4 = Register::RCX;
+        }
+        let mut set_size = |s: usize| {
+            if size_op == 0 {
+                size_op = s;
+            } else {
+                assert_eq!(size_op, s);
+            }
+        };
+        // we put the operand value into rax and than push it on stack, so the
+        // only clobbered register is rax, and if actual operand value uses it,
+        // we simply restore it from stack
+        for (op_num, op) in [op1, op2].iter().enumerate() {
+            let op_num: i64 = op_num.try_into().unwrap();
+            match op {
+                CmplogOperandType::Reg(reg) => {
+                    let info = reg.info();
+                    set_size(info.size());
+                    let reg_largest = reg.full_register();
+                    if reg_largest == Register::RAX {
+                        insts.push(
+                            Instruction::with1(
+                                Code::Push_rm64,
+                                MemoryOperand::with_base_displ(Register::RSP, op_num * 8),
+                            )
+                            .unwrap(),
+                        );
+                    } else {
+                        insts.push(Instruction::with1(Code::Push_rm64, reg_largest).unwrap());
+                    }
+                }
+                CmplogOperandType::Mem(reg_base, reg_index, disp, scale, mem_size) => {
+                    let size;
+                    let inst;
+                    match *mem_size {
+                        MemorySize::UInt64 | MemorySize::Int64 => {
+                            size = 8;
+                            inst = Code::Mov_r64_rm64;
+                        }
+                        MemorySize::UInt32 | MemorySize::Int32 => {
+                            size = 4;
+                            inst = Code::Mov_r32_rm32;
+                        }
+                        MemorySize::UInt16 | MemorySize::Int16 => {
+                            size = 2;
+                            inst = Code::Mov_r16_rm16;
+                        }
+                        _ => {
+                            println!("Invalid memory size");
+                            size = 4;
+                            inst = Code::Push_rm32;
+                        }
+                    }
+                    set_size(size);
+                    let mut disp_adjusted = *disp;
+                    if *reg_base == Register::RSP {
+                        disp_adjusted = disp_adjusted + 0x38 + 8_i64 * op_num;
```

Comment:
why 0x38?

Reply:
nevermind because we pushed 7 registers previously

Reply:
wait but are you sure rsp + 0x38 + 0x8 or rsp + 0x38 + 0x10 is free?
there might be other things in that address so this will overwrite them no?

Reply:
ehm, not sure I got you correctly.
So, we're modifying an instruction like `cmp reg, [rsp + disp]`. Since we're using the stack for our own purposes, we need to adjust `disp` accordingly, it's only going to read that memory when fetching the value.

Reply:
sorry i was confused. you are right.

Reply:
`cmp rax, dword ptr [rsp + 0x100]` is going to be transformed to (windows):
```
push 7 regs
push rax ; first cmp argument
push dword ptr [rsp + 0x100 + 7 * 8 + 8] ; second argument, 7 * 8 - saved regs, 8 - just pushed first cmd argument
mov cl, 4 ; size argument
pop rdx ; move first argument to proper register
pop r8 ; move second argument to proper register
mov r8, <address>
call populate_lists
pop 7 regs
cmp rax, dword ptr [rsp + 0x100] ; original untouched instruction
```